### PR TITLE
field: add `AddTags` method to fieldAccessor and corresponding WASM API

### PR DESCRIPTION
**Pull request by Snehil-Shah**
# adds methods to add tags to fields

- Adds `AddTags` method to pkg/datasource/accessors
- Adds corresponding methods to `Field` in WASM API (and go implementation)

## How to use

You can add tags to a field through WASM now. Example:

```go
f = ds.AddField("mntns_id", api.Kind_Uint64)
f.AddTag("type:gadget_mntns_id") // enrich mntns id
```

## Testing done

Observed implications of the above code example.

### CODE REVIEW

**Comment by alban** [comment]
The code looks good to me.

Could you add a test in pkg/operators/wasm/wasm_test.go, using `AddTag` and `HasAllTagsOf`?

**Comment by Snehil-Shah** [comment]
@alban Done!

**Review by alban — APPROVED** [review]
LGTM

File: pkg/datasource/accessors.go [hunk 1]
```diff
@@ -392,6 +395,10 @@ func (a *fieldAccessor) Tags() []string {
 	return slices.Clone(a.f.Tags)
 }
 
+func (a *fieldAccessor) AddTag(tag string) {
+	a.f.Tags = append(a.f.Tags, tag)
```
Comment (flyth) [inline pkg/datasource/accessors.go:399]: Please lock this section like done with the other public functions:

```golang
	a.ds.lock.Lock()
	defer a.ds.lock.Unlock()
```

It seems like it's missing in some other functions as well - we need to address that as well, but shouldn't introduce more functions lacking it, IMHO.

**Review by flyth — COMMENTED** [review]
Thanks for adding this!

Maybe it'd be good to have the internal API use a variadic `AddTags()` any only limit the WASM API to one tag.

**Comment by Snehil-Shah** [comment]
@flyth I didn't do that because `datasource` also has an `AddTag` method with a plural `Tags()` GET method, so a singular `AddTag` would be more consistent with the rest of the APIs, given it also has a plural `Tags()` method (I think?).

**Comment by flyth** [comment]
> @flyth I didn't do that because `datasource` also has an `AddTag` method with a plural `Tags()` GET method, so a singular `AddTag` would be more consistent with the rest of the APIs, given it also has a plural `Tags()` method (I think?).

Hmm, it seems the `AddTag()` of `datasource` is never used for now. So maybe we can instead rename that one to `AddTags()` and make it  `AddTags()`, then? WDYT?

**Comment by Snehil-Shah** [comment]
@flyth makes sense, but it will also introduce inconsistency with the wasm api's singular `AddTag` method😬.. your call.

**Comment by Snehil-Shah** [comment]
We can have `AddTags` on wasm as well, but would make the implementation a bit complicated ie. grouping bytes into groups of 8 (string enc uint64).

**Comment by alban** [comment]
The tests fail with this error message:
```
time="2024-11-04T08:55:24Z" level=warning msg="error from map iterator: bad file descriptor"
```

**Review by burak-ok — APPROVED** [review]
LGTM, thanks.
Also checked the contents of the `.tar`

**Pull request merged**